File: JRayXLib/JRayXLib/Common/Sphere.cs

/**
 * Shpere implementation
 * position is the sphere's center
 * lookAt is the direction of the north pole
 */

using JRayXLib.Math;
using JRayXLib.Math.intersections;

namespace JRayXLib.Common
{
    public class Sphere : Object3D
    {

        protected double Radius;
        protected Vect3 RotVect;

        public Sphere(Vect3 position, double radius, uint color)
            : this(position, radius, color, 0) {}

        public Sphere(Vect3 position, double radius, uint color, double reflectivity)
            : this(position, new Vect3(0, radius, 0), 0, color)
        {
            Reflectivity = reflectivity;
        }

        public Sphere(Vect3 position, Vect3 lookAt, double rotationRad, uint color, double reflectivity)
            : this(position, lookAt, rotationRad, color)
        {
            Reflectivity = reflectivity;
        }

        public Sphere(Vect3 position, Vect3 lookAt, double rotationRad, uint color) : base(position, lookAt)
        {
            Color = color;
            Radius = lookAt.Length();

            // calculate the rotation of the 0-meridian
            RotVect = new Vect3(lookAt);

            double[] rVData = RotVect.GetData();

            // check so we don't end up with two linear dependent vectors
            if (System.Math.Abs(rVData[1] - 0) > Constants.EPS
                && System.Math.Abs(rVData[0] - 0) < Constants.EPS
                && System.Math.Abs(rVData[2] - 0) < Constants.EPS)
            {
                rVData[0] += 1;
            }
            else
            {
                rVData[1] += 1;
            }
            Vect.crossProduct(RotVect, lookAt, RotVect);
            RotVect.normalize();
            LookAt.normalize();
            Rotate(lookAt, rotationRad);
        }

        public override double GetHitPointDistance(Ray r)
        {
            return RaySphere.GetHitPointRaySphereDistance(r.GetOrigin(), r.GetDirection(), Position, Radius);
        }
        
        public override void GetNormalAt(Vect3 hitPoint, Vect3 normal)
        {
            Vect.subtract(hitPoint, Position, normal);
            normal.normalize();
        }

        public override bool Contains(Vect3 hitPoint)
        {
            return System.Math.Abs(Vect.Distance(hitPoint, Position) - Radius) < Constants.EPS;
        }
        
        public override void Rotate(Matrix4 rotationMatrix)
        {
            VectMatrix.multiply(LookAt, rotationMatrix, LookAt);
            VectMatrix.multiply(RotVect, rotationMatrix, RotVect);

            LookAt.normalize();
            RotVect.normalize();
        }

        public new Sphere GetBoundingSphere()
        {
            return this;
        }

        public double GetRadius()
        {
            return Radius;
        }
    }
}

File: src/jray/common/Sphere.java
package jray.common;

import jray.math.Vect;
import jray.math.VectMatrix;
import jray.math.intersections.RaySphere;

/**
 * Shpere implementation
 * Position is the sphere's center
 * LookAt is the direction of the north pole
 */
public class Sphere extends Object3D {

    protected double radius;
    protected Vect3 rotVect;

    public Sphere(Vect3 position, double radius, int color) {
        this(position, radius, color, 0);
    }

    public Sphere(Vect3 position, double radius, int color, double reflectivity) {
        this(position, new Vect3(0, radius, 0), 0, color);
        this.reflectivity = reflectivity;
    }

    public Sphere(Vect3 position, Vect3 lookAt, double rotationRad, int color, double reflectivity) {
        this(position, lookAt, rotationRad, color);
        this.reflectivity = reflectivity;
    }

    public Sphere(Vect3 position, Vect3 lookAt, double rotationRad, int color) {
        super(position, lookAt);

        this.color = color;
        this.radius = lookAt.length();

        // calculate the rotation of the 0-meridian
        this.rotVect = new Vect3(lookAt);

        double[] rVData = rotVect.getData();

        // check so we don't end up with two linear dependent vectors
        if(rVData[1] != 0 && rVData[0] == 0 && rVData[2] == 0) {
            rVData[0] += 1;
        } else {
            rVData[1] += 1;
        }
        Vect.crossProduct(rotVect, lookAt, rotVect);
        this.rotVect.normalize();
        this.lookAt.normalize();
        this.rotate(lookAt, rotationRad);
    }

    @Override
    public double getHitPointDistance(Ray r) {
        return RaySphere.getHitPointRaySphereDistance(r.getOrigin(), r.getDirection(), position, radius);
    }

    @Override
    public void getNormalAt(Vect3 hitPoint, Vect3 normal) {
        Vect.subtract(hitPoint, position, normal);
        normal.normalize();
    }

    @Override
    public boolean contains(Vect3 hitPoint) {
        return Vect.distance(hitPoint, position) == radius;
    }

    @Override
    public void rotate(Matrix4 rotationMatrix) {
        VectMatrix.multiply(lookAt, rotationMatrix, lookAt);
        VectMatrix.multiply(rotVect, rotationMatrix, rotVect);

        lookAt.normalize();
        rotVect.normalize();
    }

    @Override
    public Sphere getBoundingSphere(){
    	return this;
    }
    
    public double getRadius() {
		return radius;
	}
}
